src/main.rs

```diff
@@ -133,27 +134,83 @@ fn main() {
 
     let pacman = alpm::Alpm::new().unwrap();
     for (pkg, cves) in infos {
-        match pacman.query_package_version(pkg.clone()) {
-            Ok(v) => {
-                info!("Found installed version {} for package {}", v, pkg);
-                for cve in cves {
-                    match cve.version {
-                        Some(version) => {
-                            info!("Comparing with fixed version {}", version);
-                            match pacman.vercmp(v.clone(), version.clone()).unwrap() {
-                                Ordering::Less => { print_asa(&options, &pkg, cve.cve, Some(version) ) },
-                                _ => {},
-                            };
-                        },
-                        None => { print_asa(&options, &pkg, cve.cve, None) },
-                    };
-                };
-            },
-            Err(_) => { debug!("Package {} not installed", pkg) },
+    	let pkg_cves = get_asas_for_package_by_version(&pacman, &pkg, &cves);
+        
+        let (cves,version) = cves_with_relevant_version(pkg_cves, &pacman);
+        
+        if !cves.is_empty() {
+
+        	print_asa(&options, &pkg, cves, version);
         }
     }
 }
 
+/// gets a map from version to ASAs for a specific package, returns list of CVEs for that
+/// package and a version, which is the most updated
+fn cves_with_relevant_version(pkg_cves: HashMap<Option<String>,Vec<ASA>>, pacman: &alpm::Alpm) -> (Vec<String>,Option<String>) {
+	let mut cves: Vec<String> = Vec::new();
+	// the newest version, which fixes all/most ? cves
+    let mut version: Option<String> = None;
+    for (v,asas) in pkg_cves {
+    	for asa in asas {
+    		for cve in asa.cve {
+    			cves.push(cve);
+    		}
+    	}
+    	match v {
+    		Some(ref v) => {
+    			version = match version {
+    				None => Some(v.clone()),
+    				Some(ref vv) => {
+    					match pacman.vercmp(v.clone(), vv.clone()).unwrap() {
+		        			Ordering::Less => Some(v.clone()),
+		        			_ => version.clone()
+		        		}
+    				}
+    			}
+    		}
+    		_ => {}
+    	}
+    }
+    (cves,version)
+}
+
+/// creates a map between version and ASAs from a list of ASAs
+fn get_asas_for_package_by_version(pacman: &alpm::Alpm, pkg: &String, cves: &Vec<ASA>) -> HashMap<Option<String>,Vec<ASA>> {
+	let mut pkg_cves: HashMap<Option<String>,Vec<ASA>> = HashMap::new();
+	match pacman.query_package_version(pkg.clone()) {
+        Ok(v) => {
+            info!("Found installed version {} for package {}", v, pkg);
+            for cve in cves {
+                match cve.version {
+                    Some(ref version) => {
+                        info!("Comparing with fixed version {}", version);
+                        match pacman.vercmp(v.clone(), version.clone()).unwrap() {
+                            Ordering::Less => { 
+                                if !pkg_cves.contains_key(&Some(v.clone())) {
```

Comment:
Could you rewrite this to match pkg_cves.get_mut() return value instead to use a if statement?